File: backend/app/api/v1/endpoints/ml.py
from fastapi import APIRouter, Query, HTTPException
from typing import Optional
from pydantic import BaseModel

from app.services.ml_service import MLService

router = APIRouter()
service = MLService()


class FarePredictionRequest(BaseModel):
    pickup_datetime: str
    pickup_latitude: float
    pickup_longitude: float
    dropoff_latitude: float
    dropoff_longitude: float
    passenger_count: int
    trip_distance: Optional[float] = None


@router.post("/predict-fare")
async def predict_fare(payload: FarePredictionRequest):
    try:
        result = await service.predict_fare(
            payload.pickup_datetime,
            payload.pickup_latitude,
            payload.pickup_longitude,
            payload.dropoff_latitude,
            payload.dropoff_longitude,
            payload.passenger_count,
            payload.trip_distance,
        )
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/model-info")
async def get_model_info():
    try:
        result = await service.get_model_info()
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/feature-importance")
async def get_feature_importance(
    model_name: str = Query("random_forest", description="Model name")
):
    try:
        result = await service.get_feature_importance(model_name)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/model-metrics")
async def get_model_metrics(
    model_name: str = Query("random_forest", description="Model name (random_forest, gbt, linear_regression)")
):
    try:
        result = await service.get_model_metrics(model_name)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: frontend/src/api/endpoints/ml.ts
import { apiClient } from '../client';

export interface ModelMetrics {
  model_type: string;
  r2: number;
  rmse: number;
  mae: number;
  training_samples: number;
  training_samples_display: string;
  feature_importance?: Record<string, number>;
}

export const getModelMetrics = async (modelName: string = 'random_forest') => {
  const response = await apiClient.get<ModelMetrics>('/ml/model-metrics', {
    params: { model_name: modelName }
  });
  return response.data;
};

export const getFeatureImportance = async (modelName: string = 'random_forest') => {
  const response = await apiClient.get('/ml/feature-importance', {
    params: { model_name: modelName }
  });
  return response.data;
};

export const predictFare = async (params: {
  pickup_datetime: string;
  pickup_latitude: number;
  pickup_longitude: number;
  dropoff_latitude: number;
  dropoff_longitude: number;
  passenger_count: number;
  trip_distance?: number;
}) => {
  const response = await apiClient.post('/ml/predict-fare', params);
  return response.data;
};

